crates/bevy_reflect/src/set.rs

```diff
@@ -0,0 +1,530 @@
+    ///
+    /// # Panics
+    ///
+    /// Panics if the given [type] is not a [`TypeInfo::Set`].
+    ///
+    /// [type]: TypeInfo
+    pub fn set_represented_type(&mut self, represented_type: Option<&'static TypeInfo>) {
+        if let Some(represented_type) = represented_type {
+            assert!(
+                matches!(represented_type, TypeInfo::Set(_)),
+                "expected TypeInfo::Set but received: {:?}",
+                represented_type
+            );
+        }
+
+        self.represented_type = represented_type;
+    }
+
+    /// Inserts a typed value into the set.
+    pub fn insert<V: Reflect>(&mut self, value: V) {
+        self.insert_boxed(Box::new(value));
+    }
+
+    #[allow(clippy::borrowed_box)]
+    fn internal_hash(value: &Box<dyn Reflect>) -> u64 {
+        value.reflect_hash().expect(hash_error!(value))
+    }
+
+    #[allow(clippy::borrowed_box)]
+    fn internal_eq(value: &Box<dyn Reflect>) -> impl FnMut(&Box<dyn Reflect>) -> bool + '_ {
+        |other| {
+            value
+                .reflect_partial_eq(&**other)
+                .expect("Not the same type")
+        }
+    }
+}
+
+// I just created this function to have only one point where we ignore the rust warning about the
+// unused allocation
+fn box_and_clone(val: &dyn Reflect) -> Box<dyn Reflect> {
+    #[allow(unused_allocation)]
+    Box::new(val).clone_value()
+}
+
+impl Set for DynamicSet {
+    fn get(&self, value: &dyn Reflect) -> Option<&dyn Reflect> {
+        let boxed = box_and_clone(value);
+        self.hash_table
+            .find(Self::internal_hash(&boxed), Self::internal_eq(&boxed))
+            .map(|value| &**value)
+    }
+
+    fn len(&self) -> usize {
+        self.hash_table.len()
+    }
+
+    fn iter(&self) -> Box<dyn Iterator<Item = &dyn Reflect> + '_> {
+        let iter = self.hash_table.iter().map(|v| &**v);
+        Box::new(iter)
+    }
+
+    fn drain(self: Box<Self>) -> Vec<Box<dyn Reflect>> {
+        self.hash_table.into_iter().collect::<Vec<_>>()
+    }
+
+    fn clone_dynamic(&self) -> DynamicSet {
+        let mut hash_table = HashTable::new();
+        self.hash_table
+            .iter()
+            .map(|value| value.clone_value())
+            .for_each(|value| {
+                hash_table.insert_unique(Self::internal_hash(&value), value, Self::internal_hash);
+            });
+
+        DynamicSet {
+            represented_type: self.represented_type,
+            hash_table,
+        }
+    }
+
+    fn insert_boxed(&mut self, value: Box<dyn Reflect>) -> bool {
+        match self
+            .hash_table
+            .find(Self::internal_hash(&value), Self::internal_eq(&value))
+        {
+            Some(_) => {
+                self.hash_table.insert_unique(
+                    Self::internal_hash(&value),
+                    value,
+                    Self::internal_hash,
+                );
+                false
+            }
```

Comment:
If this returns `Some` then the element is already in the set and it's wrong to use `insert_unique`. You should use `find_mut` and then mutate the mutable reference you get in the `Some`.